llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -3172,25 +3165,151 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+  case dwarf::DW_FORM_addrx:
+  case dwarf::DW_FORM_strx:
+  case dwarf::DW_FORM_LLVM_addrx_offset:
+  case dwarf::DW_FORM_string:
+  case dwarf::DW_FORM_indirect: {
+    return std::nullopt;
+  }
+
+  case dwarf::DW_FORM_implicit_const:
+  case dwarf::DW_FORM_flag_present: {
+    return 0;
+  }
+  case dwarf::DW_FORM_data1:
+  case dwarf::DW_FORM_ref1:
+  case dwarf::DW_FORM_flag:
+  case dwarf::DW_FORM_strx1:
+  case dwarf::DW_FORM_addrx1: {
+    return 1;
+  }
+  case dwarf::DW_FORM_data2:
+  case dwarf::DW_FORM_ref2:
+  case dwarf::DW_FORM_strx2:
+  case dwarf::DW_FORM_addrx2: {
+    return 2;
+  }
+  case dwarf::DW_FORM_strx3: {
+    return 3;
+  }
+  case dwarf::DW_FORM_data4:
+  case dwarf::DW_FORM_ref4:
+  case dwarf::DW_FORM_ref_sup4:
+  case dwarf::DW_FORM_strx4:
+  case dwarf::DW_FORM_addrx4: {
+    return 4;
+  }
+  case dwarf::DW_FORM_ref_sig8:
+  case dwarf::DW_FORM_data8:
+  case dwarf::DW_FORM_ref8:
+  case dwarf::DW_FORM_ref_sup8: {
+    return 8;
+  }
+  case dwarf::DW_FORM_data16: {
+    return 16;
+  }
+
+  case dwarf::DW_FORM_strp:
+  case dwarf::DW_FORM_sec_offset:
+  case dwarf::DW_FORM_GNU_ref_alt:
+  case dwarf::DW_FORM_GNU_strp_alt:
+  case dwarf::DW_FORM_line_strp:
+  case dwarf::DW_FORM_strp_sup: {
+    auto FormSize = FP.getDwarfOffsetByteSize();
+    return FormSize;
+  }
+  case dwarf::DW_FORM_addrx3:
+  case dwarf::DW_FORM_lo_user: {
+    llvm_unreachable("usupported form");
+    break;
+  }
+  }
+}
+
+Error DIEVisitor::materializeAbbrevDIE(unsigned AbbrevIdx) {
+  constexpr auto AddrSize = 8;
+  constexpr auto FormParams =
+      dwarf::FormParams{4 /*Version*/, AddrSize, dwarf::DwarfFormat::DWARF32};
+
   AbbrevEntryReader AbbrevReader =
       getAbbrevEntryReader(AbbrevEntries, AbbrevIdx);
 
-  if (Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag())
-    StartTagCallback(*MaybeTag, AbbrevIdx);
-  else
+  Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag();
+  if (!MaybeTag)
     return MaybeTag.takeError();
 
   Expected<bool> MaybeHasChildren = AbbrevReader.readHasChildren();
   if (!MaybeHasChildren)
     return MaybeHasChildren.takeError();
 
-  if (auto E = visitDIEAttrs(AbbrevReader, DataReader, DIEData))
+  AbbrevDIETagAndChildren TagAndChildren = {*MaybeTag, *MaybeHasChildren};
+  SmallVector<AbbrevDIEContents> AbbrevVector;
+  while (true) {
+    AbbrevDIEContents Contents;
+    Expected<dwarf::Attribute> Attr = AbbrevReader.readAttr();
+    if (!Attr)
+      return Attr.takeError();
+    if (*Attr == getEndOfAttributesMarker())
+      break;
+
+    Expected<dwarf::Form> Form = AbbrevReader.readForm();
+    if (!Form)
+      return Form.takeError();
+    Contents.Attr = *Attr;
+    Contents.Form = *Form;
+    Contents.FormInDistinctData = doesntDedup(*Form, *Attr);
+    auto FormSize = getNonULEBFormSize(*Form, FormParams);
+    Contents.FormSize = FormSize;
+    AbbrevVector.push_back(Contents);
+  }
+  AbbrevDIECache.push_back(std::make_pair(TagAndChildren, AbbrevVector));
```

Comment:
We should `std::move` this vector, since it is likely going to use memory allocations:

```
  AbbrevDIECache.emplace_back(TagAndChildren, std::move(AbbrevVector));
```

(This suggestion may vanish if you adopt my suggestion on the first commit)